apps/speed.c

```diff
@@ -2561,85 +2557,78 @@ int speed_main(int argc, char **argv)
-                                EC_KEY_get0_public_key(loopargs[i].ecdh_a[testnum]),
-                                loopargs[i].ecdh_b[testnum], loopargs[i].kdf);
-                    if (secret_size_a != secret_size_b)
-                        ecdh_checks = 0;
-                    else
-                        ecdh_checks = 1;
-
-                    for (k = 0; k < secret_size_a && ecdh_checks == 1; k++) {
-                        if (loopargs[i].secret_a[k] != loopargs[i].secret_b[k])
-                            ecdh_checks = 0;
-                    }
-
-                    if (ecdh_checks == 0) {
-                        BIO_printf(bio_err, "ECDH computations don't match.\n");
-                        ERR_print_errors(bio_err);
-                        rsa_count = 1;
-                        break;
-                    }
+                    break;
                 }
+                /* Create the context for the key generation */
+                kctx = EVP_PKEY_CTX_new(params, NULL);
+
+                EVP_PKEY_free(params); params = NULL;
+                EVP_PKEY_CTX_free(pctx); pctx = NULL;
             }
-            if (ecdh_checks != 0) {
-                pkey_print_message("", "ecdh",
-                        ecdh_c[testnum][0],
-                        test_curves_bits[testnum], ECDH_SECONDS);
-                Time_F(START);
-                count = run_benchmark(async_jobs, ECDH_compute_key_loop, loopargs);
-                d = Time_F(STOP);
-                BIO_printf(bio_err,
-                        mr ? "+R7:%ld:%d:%.2f\n" :
-                        "%ld %d-bit ECDH ops in %.2fs\n", count,
-                        test_curves_bits[testnum], d);
-                ecdh_results[testnum][0] = d / (double)count;
-                rsa_count = count;
+            if (    !kctx || /* keygen ctx is not null */
+                    !EVP_PKEY_keygen_init(kctx) || /* init keygen ctx */
+                    0) {
+                ecdh_checks = 0;
+                BIO_printf(bio_err, "ECDH keygen failure.\n");
+                ERR_print_errors(bio_err);
+                rsa_count = 1;
+                break;
             }
+
+            if (    !EVP_PKEY_keygen(kctx, &key_A) || /* generate secret key A */
+                    !EVP_PKEY_keygen(kctx, &key_B) || /* generate secret key B */
+                    !(ctx = EVP_PKEY_CTX_new(key_A, NULL)) || /* derivation ctx from skeyA */
+                    !EVP_PKEY_derive_init(ctx) || /* init derivation ctx */
+                    !EVP_PKEY_derive_set_peer(ctx, key_B) || /* set peer pubkey in ctx */
+                    0) {
+                ecdh_checks = 0;
+                BIO_printf(bio_err, "ECDH key generation failure.\n");
+                ERR_print_errors(bio_err);
+                rsa_count = 1;
+                break;
+            }
+
+            loopargs[i].ecdh_ctx[testnum] = ctx;
+
+            EVP_PKEY_CTX_free(kctx); kctx = NULL;
+        }
```

Comment:
Traditionally in speed you would have a test here to perform 1 operation. If the test passes then you go ahead with the loop of timed tests otherwise you fail. The actually timing loop is then not required to check for errors (for maximum efficiency) and is assumed to be running correctly. This isn't done consistently throughout speed as different people have done different things as they have added new tests. Previously for ECDH there were 2 ECDH_compute_key operations performed and compared for length as test before running the timed loop. This appears to have not been converted in the new code, was this intentional?